### ISSUE TITLE
Unexpected behavior of torch.pow with both tensor args and nested jacfwd or jacrev

### ISSUE THREAD

**Issue by i-a-morozov** [pytorch/pytorch#89757]
Hi, `torch.pow` with two tensor arguments results in wrong higher order derivatives.

```python

# functorch.__version__
# '1.13.0+cu117'

import torch
import functorch

jacobian = functorch.jacrev

x = torch.tensor([0.0, 0.0], dtype=torch.float64)
y = torch.tensor([0.0, 0.0], dtype=torch.float64)

def foo(x, y):
    x1, x2 = x
    y1, y2 = y
    return torch.stack([x1*x2*y1*y2])

def bar1(x, y):
    power = torch.tensor([1, 1, 1, 1], dtype=torch.int64)
    state = torch.cat([x, y])
    state = torch.pow(state, power)
    return torch.stack([state.prod(-1)])

def bar2(x, y):
    power = [1, 1, 1, 1]
    # power = torch.tensor([1, 1, 1, 1], dtype=torch.int64) # -- will also break
    state = torch.cat([x, y])
    state = torch.stack([s**p for s, p in zip(state, power)])
    return torch.stack([state.prod(-1)])

print(foo(x, y).numpy())
print(bar1(x, y).numpy())
print(bar2(x, y).numpy())

# d^2(d^2(f)/d(y^2))/d(x^2)

print(jacobian(jacobian(lambda x: jacobian(jacobian(lambda y: foo(x, y)))(y)))(x).flatten().numpy())
print(jacobian(jacobian(lambda x: jacobian(jacobian(lambda y: bar1(x, y)))(y)))(x).flatten().numpy())
print(jacobian(jacobian(lambda x: jacobian(jacobian(lambda y: bar2(x, y)))(y)))(x).flatten().numpy())

# jacobian = functorch.jacrev
# [0.]
# [0.]
# [0.]
# [0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0.]
# [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan]
# [0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0.]

# jacobian = functorch.jacfwd
# [0.]
# [0.]
# [0.]
# [0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0.]
# [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
# [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

```

**Comment by zou3519** [comment]
Moving this to the pytorch/pytorch repo to better track it.

Thanks for the bug report, @i-a-morozov. I'm not sure what is going on yet, but I agree that it shouldn't be outputting nans. 

It looks like there is some incorrect interaction between cumprod/prod backward and the handling of multiple zeros being multiplied together. cc @kshitij12345 -- this reminds me of some of the composite compliance work there.

**Comment by zou3519** [comment]
I am able to reproduce this on PyTorch 1.12.1 and functorch 0.2.1, so it is not technically a regression

**Comment by atalman** [comment]
Since its not regression removing from 1.13.1 milestone

**Comment by albanD** [comment]
Tentative for 1.13.1. But we won't delay the release on this if it is not fixed by the time we do the branch cut for it.

**Comment by atalman** [comment]
hi @zou3519 could you please let us know ETA for this issue.
Following are the tentative dates for 1.13.1:
Phase 1 (until 12/8/2022): Release Candidate 1
Phase 2 (until 12/15/2022): Tentative General Availability

**Comment by kshitij12345** [comment]
@zou3519 ,

This seems to be unrelated to `prod`/`cumprod` (replacing them with `sum`) also leads to the same error. 

I believe the problem is with `pow`. Below is a small reproducer with `pow`.
```python
import torch
import functorch
from functorch import make_fx

jacobian = functorch.jacrev

def fn(x):
    exp = torch.tensor([1.], dtype=torch.double)
    # exp = 1 # This works
    return torch.pow(x, exp)

x = torch.tensor([0.0,], dtype=torch.float64)
o = jacobian(jacobian(jacobian(fn)))(x)
print(o)
```

Output
```
tensor([[[[0.]]]], dtype=torch.float64)  # exp = 1 i.e. TensorXScalar overload
tensor([[[[nan]]]], dtype=torch.float64)  # exp = torch.tensor([1.], dtype=torch.double) i.e. TensorXTensor overload
```

**Comment by zou3519** [comment]
Thanks for the analysis, @kshitij12345 

I dug a bit more, it looks like this isn't a functorch-specific problem. Our gradients for torch.pow incorrectly return nan at the third-order:
```py
import torch
import functorch
from functorch import make_fx

jacobian = functorch.jacrev

def fn(x):
    exp = torch.tensor(1., dtype=torch.double)
    # exp = 1 # This works
    return torch.pow(x, exp)

x = torch.tensor(0.0, dtype=torch.float64, requires_grad=True)

gx, = torch.autograd.grad(fn(x), x, create_graph=True)
ggx, = torch.autograd.grad(gx, x, create_graph=True)
gggx, = torch.autograd.grad(ggx, x)
print(gx, ggx, gggx)
```

**Comment by kshitij12345** [comment]
There is special handling `exponent=0` case which seems to work fine till 2nd order.

https://github.com/pytorch/pytorch/blob/a6caa9c54bb5fce93a77e7407fd7e8993a767527/torch/csrc/autograd/FunctionsManual.cpp#L426-L430

**Comment by zou3519** [comment]
Given that this problem has been around for a while, I'm removing the priority on this and removing it from the milestone for now and deferring to the autograd folks to prioritize

**Comment by juanf-0gravity** [comment]
Ran into this while working with higher-order gradients and took a look.

As @zou3519 and @kshitij12345 pointed out, the problem is in `pow_backward_self` (`FunctionsManual.cpp`). The `at::where` guards against `exponent == 0`, but `self.pow(exponent - 1)` still gets eagerly evaluated. At 3rd order with `pow(0., 1.)`, the backward chain hits `0^(-1) = inf`, `MulBackward0` saves that, and next backward pass does `0 * inf = NaN`.

I noticed `i1_backward` (around line 5819) handles a similar situation by swapping in safe values before computing so the dead branch stays finite:

```cpp
auto safe_self = at::where(cond, at::ones_like(self), self);
```

I think the same approach could work here. Swap `self` with ones when `exponent == 0` so `pow` always gets a finite base, then let `at::where` pick the right result.

Would a PR for this be welcome? Happy to take a shot at it.
